File: services/py/src/application/limiter/bucket.py
import time
from asyncio import Lock


class TokenBucket:
    def __init__(self, rate: int, capacity: int) -> None:
        self.rate = rate
        self.capacity = capacity
        self.tokens = capacity
        self.timestamp = time.monotonic()
        self._lock = Lock()

    async def consume(self, tokens: int = 1) -> bool:
        async with self._lock:
            now = time.monotonic()
            elapsed = now - self.timestamp
            self.timestamp = now

            self.tokens = min(self.capacity, self.tokens + elapsed * self.rate)
            if self.tokens >= tokens:
                self.tokens -= tokens
                return True

            return False

File: services/go/src/internal/application/limiter/bucket.go
package limiter

import (
	"sync"
	"time"
)

type TokenBucket struct {
	mu       sync.RWMutex
	tokens   float64
	last     time.Time
	rate     float64
	capacity float64
}

func NewTokenBucket(rate, capacity float64) *TokenBucket {
	return &TokenBucket{
		tokens:   capacity,
		last:     time.Now(),
		rate:     rate,
		capacity: capacity,
	}
}

func (b *TokenBucket) Allow() bool {
	b.mu.Lock()
	defer b.mu.Unlock()

	elapsed := time.Since(b.last).Seconds()
	b.tokens += elapsed * b.rate
	if b.tokens > b.capacity {
		b.tokens = b.capacity
	}
	b.last = time.Now()

	if b.tokens < 1 {
		return false
	}

	b.tokens -= 1
	return true
}
